spec/firebase_app_distribution_action_spec.rb

```diff
@@ -1,131 +1,193 @@
 describe Fastlane::Actions::FirebaseAppDistributionAction do
   let(:fake_file) { StringIO.new }
-  let(:fake_connection) { double("Connection") }
   let(:fake_binary) { double("Binary") }
+  let(:fake_auth_client) { double("auth_client") }
+  let(:stubs) { Faraday::Adapter::Test::Stubs.new }
+  let(:conn) do
+    Faraday.new(url: "https://firebaseappdistribution.googleapis.com") do |b|
+      b.response(:json, parser_options: { symbolize_names: true })
+      b.response(:raise_error)
+      b.adapter(:test, stubs)
+    end
+  end
 
   before(:each) do
-    allow(Fastlane::Actions::FirebaseAppDistributionAction).to receive(:connection).and_return(fake_connection)
+    allow(Signet::OAuth2::Client).to receive(:new).and_return(fake_auth_client)
+    allow(fake_auth_client).to receive(:fetch_access_token!)
+    allow(fake_auth_client).to receive(:access_token).and_return("fake_auth_token")
+
     allow(fake_binary).to receive(:read).and_return("Hello World")
     allow(File).to receive(:open).and_return(fake_binary)
+
+    allow(Fastlane::Actions::FirebaseAppDistributionAction).to receive(:connection).and_return(conn)
+  end
+
+  after(:each) do
+    stubs.verify_stubbed_calls
+    Faraday.default_connection = nil
   end
 
   describe '#get_upload_token' do
-    context 'when testing with valid parameters' do
-      it 'should make a GET call to the app endpoint and return the upload token' do
-        expect(fake_connection).to receive(:get)
-          .with("/v1alpha/apps/app_id")
-          .and_return(
-            double("Response", status: 200, body: {
-              projectNumber: "project_number",
-              appId: "app_id",
-              platform: "android",
-              bundleId: "bundle_id",
-              contactEmail: "Hello@world.com"
-            })
-          )
-        upload_token = Fastlane::Actions::FirebaseAppDistributionAction.get_upload_token("app_id", "binary_path")
-        binary_hash = Digest::SHA256.hexdigest("Hello World")
-        expect(upload_token).to eq(CGI.escape("projects/project_number/apps/app_id/releases/-/binaries/#{binary_hash}"))
+    it 'should make a GET call to the app endpoint and return the upload token' do
+      stubs.get("/v1alpha/apps/app_id") do |env|
+        expect(env.url.path).to eq("/v1alpha/apps/app_id")
```

Comment:
This stubs thing is very cool. I'm curious: what value do you get from this expect? Won't this always be true, because if the url path was different, this block wouldn't get called?

Also, could we use `stubs.verify_stubbed_calls` to verify that the stub was called with the given url and body? See https://github.com/lostisland/faraday/blob/master/docs/adapters/testing.md

Reply:
In this instance this will always be true. However, in cases like this: 

https://github.com/fastlane/fastlane-plugin-firebase_app_distribution/blob/945ec172480adb3b69f96018e5aa3fe3b2689ac2/spec/firebase_app_distribution_action_spec.rb#L90-L91

The env.url.path will be different than the stubs.post call because it does not include the additional args, only the path.

Reply:
Also we do use the stubs.verify_stubbed_calls here: 

https://github.com/fastlane/fastlane-plugin-firebase_app_distribution/blob/945ec172480adb3b69f96018e5aa3fe3b2689ac2/spec/firebase_app_distribution_action_spec.rb#L25-L28

Reply:
Ah! I missed the after block. Thanks.

Even in the case like `/app-binary-uploads` I don't see the value in expecting the URL path. Even though it's different from the stubs.post call, it will still always be true because the block wouldn't execute otherwise.